src/main/java/org/spongepowered/common/event/tracking/PhaseStack.java

```diff
@@ -133,19 +133,18 @@ boolean checkForRunaways(IPhaseState<?> state, @Nullable PhaseContext<?> phaseCo
             return false;
         }
         final int totalCount = this.phases.size();
-        final PhaseContext<?>[] allContexts = new PhaseContext[totalCount];
-        int i = 0;
+        if (totalCount < 2) {
+            return false;
+        }
         // So first, we want to collect all the states into an array as they are pushed to the stack,
         // which means that we should see the re-entrant phase pretty soon.
-        for (PhaseContext<?> data : this.phases) {
-            allContexts[i++] = data;
-        }
+        final Object[] allContexts = this.phases.toArray();
```

Comment:
Just use a system arraycopy on a properly typed array, you add an extra checklist later on.

Reply:
I need the source array to call `System#arraycopy` directly. The source array is a protected field (`elements`) inside the `ArrayDeque` class, so I can only use `Deque#toArray();`.

`toArray()` is [much faster](https://shipilev.net/blog/2016/arrays-wisdom-ancients/) than the type-checked arraycopy of `toArray(T[] a)` and the `checkcast` would be redundant only if I used the latter.